src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -794,9 +794,11 @@ public void shutdown() {
         QNode headNext;
         for (;;) {
             headNext = head.getNext();
-            if (headNext instanceof TaskNode) {
+            if (headNext != head && headNext instanceof TaskNode) {
                 TaskNode taskNode = (TaskNode) headNext;
                 if (compareAndSetHead(head, taskNode)) {
+                    // save from GC nepotism
+                    head.setNextRelaxed(head);
```

Comment:
Can you prove that this can never cause nodes to be skipped?

Reply:
good point: looking at the original code I see that is draining `TaskNode`s until `tail::getNext` is null, after a shutdown: I can preserve the original logic to handle the case of a concurrent call to it.

Reply:
I've changed the visibility guarantess while setting next too, taking "inspiration" from CLQ:
https://github.com/openjdk-mirror/jdk7u-jdk/blob/8aaca72128077594ecf4ecdee761e78d4ecb8b7d/src/share/classes/java/util/concurrent/ConcurrentLinkedQueue.java#L305

```java
    final void updateHead(Node<E> h, Node<E> p) {
        if (h != p && casHead(h, p))
            h.lazySetNext(h);
    }
```